optimum/intel/openvino/trainer.py

```diff
@@ -0,0 +1,558 @@
+from tqdm.auto import tqdm
+from transformers import Trainer
+from transformers.data.data_collator import DataCollator
+from transformers.debug_utils import DebugOption, DebugUnderflowOverflow
+from transformers.integrations import hp_params
+from transformers.modeling_utils import PreTrainedModel, unwrap_model
+from transformers.onnx import FeaturesManager, OnnxConfig
+from transformers.tokenization_utils_base import PreTrainedTokenizerBase
+from transformers.trainer import TRAINER_STATE_NAME, TRAINING_ARGS_NAME
+from transformers.trainer_callback import TrainerCallback, TrainerState
+from transformers.trainer_pt_utils import IterableDatasetShard
+from transformers.trainer_utils import EvalPrediction, HPSearchBackend, TrainOutput, has_length, speed_metrics
+from transformers.training_args import TrainingArguments
+from transformers.utils import WEIGHTS_NAME, TensorType, is_sagemaker_mp_enabled, logging
+
+import openvino
+from nncf import NNCFConfig
+from nncf.config.structures import BNAdaptationInitArgs, QuantizationRangeInitArgs
+from nncf.torch import create_compressed_model
+from nncf.torch.nncf_network import NNCFNetwork
+from openvino.offline_transformations import compress_quantize_weights_transformation
+from openvino.runtime import Core
+from optimum.utils import logging
+
+from .configuration import OVConfig
+from .quantization import OVDataLoader
+from .utils import OV_XML_FILE_NAME
+
+
+core = Core()
+
+logger = logging.get_logger(__name__)
+
+
+class OVTrainer(Trainer):
+    """
+    OVTrainer enables NNCF quantization aware training.
+    """
+
+    def __init__(
+        self,
+        model: Union[PreTrainedModel, torch.nn.Module] = None,
+        args: TrainingArguments = None,
+        data_collator: Optional[DataCollator] = None,
+        train_dataset: Optional[Dataset] = None,
+        eval_dataset: Optional[Dataset] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        model_init: Callable[[], PreTrainedModel] = None,
+        compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
+        callbacks: Optional[List[TrainerCallback]] = None,
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        preprocess_logits_for_metrics: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] = None,
+        ov_config: Optional[OVConfig] = None,
+        feature: Optional[str] = None,
+    ):
+
+        super().__init__(
+            model,
+            args,
+            data_collator,
+            train_dataset,
+            eval_dataset,
+            tokenizer,
+            model_init,
+            compute_metrics,
+            callbacks,
+            optimizers,
+            preprocess_logits_for_metrics,
+        )
+
+        self.ov_config = ov_config
+        self.feature = feature
+        self.compression_controller = None
+
+        if self.ov_config is not None and self.args.do_train:
```

Comment:
Let's do the same for `ov_config` in case it is `None` as we do it in `OVQuantizer`.

Reply:
For the `OVQuantizer`, it make sense to me that the user will always want to apply static quantization when using it (so the `OVConfig` can be kept optional). However, for the `OVTrainer`, a user could be interested in training a model with and without QAT in order to compare both, so I am in favor of keeping QAT optional (enabling it when an `OVConfig` is given).

Reply:
OV specific is only in QAT, without it it does not make sense to replace `Trainer` with `OVTrainer`. So, you can use `Trainer` and then use `OVModelFor...`

Reply:
What if we integrate additional compression techniques in the future such as pruning or distillation, QAT shouldn't be the default option in my opinion. Also having a config specifying the quantization process would keep consistency with `optimum.intel.neural_compressor.IncTrainer`.

Reply:
Ok, let's go with the explicit definition of the config file.